pkg/observer/polling/observer.go

```diff
@@ -0,0 +1,337 @@
+		ratio:            ratio,
+		stager:           &stager{},
+		coordinator:      coord,
+		src:              src,
+		runner:           runner,
+		encoder:          encoder,
+		heads:            heads,
+		mercuryLookup:    mercuryLookup,
+	}
+
+	// make all go-routines started by this entity automatically recoverable
+	// on panics
+	ob.services = []Service{
+		util.NewRecoverableService(&observer.SimpleService{F: ob.runHeadTasks, C: cancel}, logger),
+	}
+
+	return ob
+}
+
+type PollingObserver struct {
+	ctx       context.Context
+	cancel    context.CancelFunc
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// static values provided by constructor
+	samplingDuration time.Duration // limits time spent processing a single block
+	ratio            Ratio         // ratio for limiting sample size
+
+	// initialized components inside a constructor
+	services []Service
+	stager   *stager
+
+	// dependency interfaces required by the polling observer
+	logger      *log.Logger
+	heads       HeadProvider                    // provides new blocks to be operated on
+	coordinator ocr2keepers.Coordinator         // key status coordinator tracks in-flight status
+	shuffler    Shuffler[ocr2keepers.UpkeepKey] // provides shuffling logic for upkeep keys
+
+	src           UpkeepProvider
+	runner        Runner
+	encoder       Encoder
+	mercuryLookup bool
+}
+
+// Observe implements the Observer interface and provides a slice of identifiers
+// that were observed to be performable along with the block at which they were
+// observed. All ids that are pending are filtered out.
+func (o *PollingObserver) Observe() (ocr2keepers.BlockKey, []ocr2keepers.UpkeepIdentifier, error) {
+	bl, ids := o.stager.get()
+	filteredIDs := make([]ocr2keepers.UpkeepIdentifier, 0, len(ids))
+
+	for _, id := range ids {
+		key := o.encoder.MakeUpkeepKey(bl, id)
+
+		if pending, err := o.coordinator.IsPending(key); pending || err != nil {
+			if err != nil {
+				o.logger.Printf("error checking pending state for '%s': %s", key, err)
+			} else {
+				o.logger.Printf("filtered out key '%s'", key)
+			}
+
+			continue
+		}
+
+		filteredIDs = append(filteredIDs, id)
+	}
+
+	return bl, filteredIDs, nil
+}
+
+// Start will start all required internal services. Calling this function again
+// after the first is a noop.
+func (o *PollingObserver) Start() {
+	o.startOnce.Do(func() {
+		for _, svc := range o.services {
+			o.logger.Printf("PollingObserver service started")
+
+			svc.Start()
+		}
+	})
+}
+
+// Stop will stop all internal services allowing the observer to exit cleanly.
+func (o *PollingObserver) Close() error {
+	o.stopOnce.Do(func() {
+		for _, svc := range o.services {
+			o.logger.Printf("PollingObserver service stopped")
+
+			svc.Stop()
+		}
+	})
+
+	return nil
+}
+
+func (o *PollingObserver) runHeadTasks() error {
+	ch := o.heads.HeadTicker()
+	for {
+		select {
+		case bl := <-ch:
```

Comment:
As I understand correctly, it expected that this channel will drop heads while the current one is being processed?

I feel this is a very non trivial assumption and would be good to surface that our more clearly (comments / renaming `HeadTicker` variable to make this clear